File: cpp/example_code/cloudwatch/cw_put_metric_alarm.cpp
/*
   Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.

   This file is licensed under the Apache License, Version 2.0 (the "License").
   You may not use this file except in compliance with the License. A copy of
   the License is located at

    http://aws.amazon.com/apache2.0/

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
   CONDITIONS OF ANY KIND, either express or implied. See the License for the
   specific language governing permissions and limitations under the License.
*/
#include <aws/core/Aws.h>

#include <aws/monitoring/CloudWatchClient.h>
#include <aws/monitoring/model/PutMetricAlarmRequest.h>

#include <iostream>

/**
 * Adds a cloud watch alarm to an instance based on CPU utilization
 */
int main(int argc, char** argv)
{
    if (argc != 3)
    {
        std::cout << "Usage: cw_put_metric_alarm <alarm_name> <instance_id>" << std::endl;
        return 1;
    }

    Aws::String alarmName(argv[1]);
    Aws::String instanceId(argv[2]);

    Aws::SDKOptions options;
    Aws::InitAPI(options);

    {
        Aws::CloudWatch::CloudWatchClient cw_client;

        Aws::CloudWatch::Model::PutMetricAlarmRequest putMetricAlarmRequest;
        putMetricAlarmRequest.SetAlarmName(alarmName);
        putMetricAlarmRequest.SetComparisonOperator(Aws::CloudWatch::Model::ComparisonOperator::GreaterThanThreshold);
        putMetricAlarmRequest.SetEvaluationPeriods(1);
        putMetricAlarmRequest.SetMetricName("CPUUtilization");
        putMetricAlarmRequest.SetNamespace("AWS/EC2");
        putMetricAlarmRequest.SetPeriod(60);
        putMetricAlarmRequest.SetStatistic(Aws::CloudWatch::Model::Statistic::Average);
        putMetricAlarmRequest.SetThreshold(70.0);
        putMetricAlarmRequest.SetActionsEnabled(false);
        putMetricAlarmRequest.SetAlarmDescription("Alarm when server CPU exceeds 70%");
        putMetricAlarmRequest.SetUnit(Aws::CloudWatch::Model::StandardUnit::Seconds);

        Aws::CloudWatch::Model::Dimension dimension;
        dimension.SetName("InstanceId");
        dimension.SetValue(instanceId);

        putMetricAlarmRequest.AddDimensions(dimension);

        auto putMetricAlarmOutcome = cw_client.PutMetricAlarm(putMetricAlarmRequest);
        if (!putMetricAlarmOutcome.IsSuccess())
        {
            std::cout << "Failed to create cloudwatch alarm:" << putMetricAlarmOutcome.GetError().GetMessage() <<
            std::endl;
        }
        else
        {
            std::cout << "Successfully created cloudwatch alarm " << alarmName << std::endl;
        }
    }

    Aws::ShutdownAPI(options);

    return 0;
}

File: javascript/example_code/cloudwatch/cw_putmetricalarm.js
/**
 * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * This file is licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License. A copy of
 * the License is located at
 *
 * http://aws.amazon.com/apache2.0/
 *
 * This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
*/

//snippet-sourcedescription:[cw_putmetricalarm.js demonstrates how to create or update an Amazon CloudWatch alarm and associate it with particular metrics.]
//snippet-service:[cloudwatch]
//snippet-keyword:[JavaScript]
//snippet-keyword:[Code Sample]
//snippet-keyword:[Amazon CloudWatch]
//snippet-sourcetype:[full-example]
//snippet-sourcedate:[2018-06-02]
//snippet-sourceauthor:[AWS-JSDG]

// ABOUT THIS NODE.JS SAMPLE: This sample is part of the SDK for JavaScript Developer Guide topic at
// https://docs.aws.amazon.com/sdk-for-javascript/v2/developer-guide/cloudwatch-examples-creating-alarms.html 

// snippet-start:[cw.JavaScript.alarms.putMetricAlarm]
// Load the AWS SDK for Node.js
var AWS = require('aws-sdk');
// Set the region 
AWS.config.update({region: 'REGION'});

// Create CloudWatch service object
var cw = new AWS.CloudWatch({apiVersion: '2010-08-01'});

var params = {
  AlarmName: 'Web_Server_CPU_Utilization',
  ComparisonOperator: 'GreaterThanThreshold',
  EvaluationPeriods: 1,
  MetricName: 'CPUUtilization',
  Namespace: 'AWS/EC2',
  Period: 60,
  Statistic: 'Average',
  Threshold: 70.0,
  ActionsEnabled: false,
  AlarmDescription: 'Alarm when server CPU exceeds 70%',
  Dimensions: [
    {
      Name: 'InstanceId',
      Value: 'INSTANCE_ID'
    },
  ],
  Unit: 'Percent'
};

cw.putMetricAlarm(params, function(err, data) {
  if (err) {
    console.log("Error", err);
  } else {
    console.log("Success", data);
  }
});
// snippet-end:[cw.JavaScript.alarms.putMetricAlarm]
